crypto/representationproofs/representation.go

```diff
@@ -0,0 +1,148 @@
+	for i := 0; i < 3; i++ {
+		r := common.GetRandomInt(group.Q)
+		base := group.Exp(group.G, r)
+		bases = append(bases, base)
+	}
+
+	var secrets []*big.Int
+	for i := 0; i < 3; i++ {
+		secret := common.GetRandomInt(group.Q)
+		secrets = append(secrets, secret)
+	}
+
+	// y = g_1^x_1 * ... * g_k^x_k where g_i are bases and x_i are secrets
+	y := big.NewInt(1)
+	for i := 0; i < 3; i++ {
+		f := group.Exp(bases[i], secrets[i])
+		y = group.Mul(y, f)
+	}
+
+	prover, err := NewRepresentationProver(group, secrets, bases, y)
+	if err != nil {
+		fmt.Printf("Error when instantiating RepresentationProver")
+		return false
+	}
+	verifier := NewRepresentationVerifier(group, bases, y)
+
+	proofRandomData := prover.GetProofRandomData()
+	verifier.SetProofRandomData(proofRandomData)
+
+	challenge := verifier.GetChallenge()
+	proofData := prover.GetProofData(challenge)
+	verified := verifier.Verify(proofData)
+	return verified
+}
+
+// Proving that it knows (x_1,...,x_k) such that y = g_1^x_1 * ... * g_k^x_k where g_i are
+// given generators of cyclic group G. Note that Schnorr is a special case of RepresentationProver
+// where only one base is used.
+type RepresentationProver struct {
+	Group            *groups.ModCyclicGroup
+	secrets          []*big.Int
+	bases            []*big.Int
+	randomValues     []*big.Int
+	y 				 *big.Int
+}
+
+func NewRepresentationProver(group *groups.ModCyclicGroup, secrets,
+		bases []*big.Int, y *big.Int) (*RepresentationProver, error) {
+	if len(secrets) != len(bases) {
+		return nil, fmt.Errorf("Number of secrets and representation bases shoud be the same.")
+	}
+
+	var prover RepresentationProver
+	prover = RepresentationProver{
+		Group:         group,
+		secrets:	   secrets,
+		bases: 		   bases,
+		y: 			   y,
+	}
+
+	return &prover, nil
+}
+
+func (prover *RepresentationProver) GetProofRandomData() *big.Int {
+	// t = g_1^r_1 * ... * g_k^r_k where g_i are bases and r_i are random values
+	t := big.NewInt(1)
+	var randomValues []*big.Int
+	for i := 0; i < len(prover.bases); i++ {
+		r := common.GetRandomInt(prover.Group.Q)
+		randomValues = append(randomValues, r)
+		f := prover.Group.Exp(prover.bases[i], r)
+		t = prover.Group.Mul(t, f)
+	}
+	prover.randomValues = randomValues
+	return t
+}
+
+func (prover *RepresentationProver) GetProofData(challenge *big.Int) []*big.Int {
+	// z_i = r_i + challenge * secrets[i]
+	var proofData []*big.Int
+	for i := 0; i < len(prover.bases); i++ {
+		z_i := prover.Group.Mul(challenge, prover.secrets[i])
+		z_i = prover.Group.Add(z_i, prover.randomValues[i])
+		proofData = append(proofData, z_i)
+	}
+	return proofData
+}
+
+type RepresentationVerifier struct {
+	Group             *groups.ModCyclicGroup
+	bases            []*big.Int
+	proofRandomData  *big.Int
+	y 				 *big.Int
+	challenge         *big.Int
+}
+
+func NewRepresentationVerifier(group *groups.ModCyclicGroup, bases []*big.Int,
+		y *big.Int) *RepresentationVerifier {
+	verifier := RepresentationVerifier{
```

Comment:
Please modify L116-L121 so that you just `return &RepresentationVerifier{...}`, and fix indentation.